File: apple/Sources/SwiftyKit/Padding.swift
import CaladonCoreFFI
import Foundation

/// Metadata padding — the sealed-sender SIZE analog (identity-envelope.md §9). Pad the
/// plaintext to a fixed bucket before sealing so the ciphertext length reveals only the
/// bucket, not the true size. Now a thin wrapper over the shared `caladon-core` Rust trust
/// core (`CaladonCoreFFI.pad`/`unpad`), which is byte-identical to the Python reference
/// `swifty_crypto.padding` (deterministic: uint32_be(len) ‖ plaintext ‖ zero filler → next
/// bucket) and to the prior in-Swift implementation.
public enum Padding {
    public static let buckets: [Int] = [256, 1024, 4096, 16384, 65536, 262144]
    static let lenPrefix = 4

    public enum PaddingError: Error, Equatable { case tooShort, badLength }

    /// Pad to a fixed bucket. `pad(x).count` reveals only the bucket, not `x.count`.
    /// The Rust core's only failure is a plaintext longer than 2^32 bytes (impossible for any
    /// real payload), so this keeps the historically non-throwing surface.
    public static func pad(_ plaintext: Data) -> Data {
        // `pad` over the FFI throws only `PadError.TooLong` (len > u32::MAX), unreachable here.
        try! CaladonCoreFFI.pad(plaintext: plaintext)
    }

    /// Recover the exact plaintext. Fail-closed on a malformed/short buffer.
    public static func unpad(_ padded: Data) throws -> Data {
        do {
            return try CaladonCoreFFI.unpad(padded: padded)
        } catch CaladonCoreFFI.PadError.TooShort {
            throw PaddingError.tooShort
        } catch {
            // DeclaredLenExceedsBuffer (or any other malformed-buffer failure) → fail closed.
            throw PaddingError.badLength
        }
    }
}

File: caladon-core/src/padding.rs
//! Metadata padding — the sealed-sender SIZE analog (identity-envelope.md §9). Byte-identical to
//! `swifty_crypto/padding.py`: `uint32_be(len) ‖ plaintext ‖ zero-fill` to the next bucket.

const PAD_BUCKETS: [usize; 6] = [256, 1024, 4096, 16384, 65536, 262144];
const LEN_PREFIX: usize = 4;
const MAX_LEN: usize = 0xFFFF_FFFF;

#[derive(Debug, PartialEq, Eq)]
pub enum PadError {
    TooLong,
    TooShort,
    DeclaredLenExceedsBuffer,
}

fn bucket(n: usize) -> usize {
    for &b in PAD_BUCKETS.iter() {
        if n <= b {
            return b;
        }
    }
    let largest = PAD_BUCKETS[PAD_BUCKETS.len() - 1];
    ((n + largest - 1) / largest) * largest // round up to a multiple of the largest
}

/// Pad to a fixed bucket. `pad(x).len()` reveals only the bucket, not `x.len()`.
pub fn pad(plaintext: &[u8]) -> Result<Vec<u8>, PadError> {
    if plaintext.len() > MAX_LEN {
        return Err(PadError::TooLong);
    }
    let mut body = Vec::with_capacity(LEN_PREFIX + plaintext.len());
    body.extend_from_slice(&(plaintext.len() as u32).to_be_bytes());
    body.extend_from_slice(plaintext);
    let target = bucket(body.len());
    body.resize(target, 0u8);
    Ok(body)
}

/// Recover the exact plaintext. Fail-closed on a malformed/short buffer.
pub fn unpad(padded: &[u8]) -> Result<Vec<u8>, PadError> {
    if padded.len() < LEN_PREFIX {
        return Err(PadError::TooShort);
    }
    let n = u32::from_be_bytes([padded[0], padded[1], padded[2], padded[3]]) as usize;
    if n > padded.len() - LEN_PREFIX {
        return Err(PadError::DeclaredLenExceedsBuffer);
    }
    Ok(padded[LEN_PREFIX..LEN_PREFIX + n].to_vec())
}
